src/raft/raft.rs

```diff
@@ -916,15 +938,46 @@ impl<T: Storage> Raft<T> {
                 self.handle_append_response(m, old_paused, send_append, maybe_commit);
             }
             MessageType::MsgHeartbeatResponse => {
-                let pr = self.prs.get_mut(&m.get_from()).unwrap();
-                pr.recent_active = true;
+                {
+                    let pr = self.prs.get_mut(&m.get_from()).unwrap();
+                    pr.recent_active = true;
+
+                    // free one slot for the full inflights window to allow progress.
+                    if pr.state == ProgressState::Replicate && pr.ins.full() {
+                        pr.ins.free_first_one();
+                    }
+                    if pr.matched < self.raft_log.last_index() {
+                        *send_append = true;
```

Comment:
I don't get the point why to use flag instead of calling send_append() directly like etcd/raft. But it's certain the read index heartbeats don't matter the normal append entries message.